Automattic/vip-go-mu-plugins
# Search: Add base index versioning support

## Description

Adds the foundational index versioning code for tracking and managing versions.

Versions are stored inside an option (or a site option if in network mode), which has the following format:

```
array(
    'post' => array(
        array(
            'number' => 2,
            'active' => true,
            'created_time => 123456789,
            'activated_time' => 123456789,
        ),
        array(
            'number' => 3,
            'active' => false,
            'created_time => 123456789,
        )
    ),
    'user' => array(

    ),
    ...
)
```

This PR adds basic management of index versions and adds the filtering to the index name so that when versions are activated, the new index is used by default. For existing indexes (without any versions), the version is always 1, and we do _not_ add v1 to the index name (for backwards compatibility).

Previous index names:

```
vip-123-post-1
```

New index names (for versions > 2):

```
vip-123-post-1-v2
```

To run queries against non-active indexes (for example, to build up a new index in-place while another is active), simply override the current index number with:

```
$search->versioning->set_current_index_number( $indexable, 42 );
```

From then on, all queries will go to index version `42`, until the index number is overridden again, reset via `reset_current_index_number()`, or the request completes.

## Checklist

Please make sure the items below have been covered before requesting a review:


## Steps to Test

1. Setup a site using VIP Search prior to switching to PR - make sure index exists
1. Pull down PR
1. Perform some queries - everything should work the same as before
1. Re-index the content - everything should be the same (and index name is like `vip-123-post-1` (no version)
1. Create a new version...in wp shell: 
```
$indexable = \ElasticPress\Indexables::factory()->get( 'post' );
\Automattic\VIP\Search\Search::instance()->versioning->add_version( $indexable );
\Automattic\VIP\Search\Search::instance()->versioning->activate_version( $indexable, 2 );
```
1. Re-index content normally - `wp vip-search index --setup`
1. Perform some new queries - should now be hitting v2 of the index - `vip-123-post-1-v2`
1. Try to activate a non-existent index: 
```
$indexable = \ElasticPress\Indexables::factory()->get( 'post' );
\Automattic\VIP\Search\Search::instance()->versioning->activate_version( $indexable, 3 );
```
1. That should fail with an error b/c that index hasn't yet been registered via `add_version`
